File: android/beagle/src/main/java/br/com/zup/beagle/android/components/page/PageIndicator.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.android.components.page

import android.graphics.Color
import br.com.zup.beagle.android.context.Bind
import br.com.zup.beagle.android.context.valueOf
import br.com.zup.beagle.android.utils.observeBindChanges
import br.com.zup.beagle.android.view.ViewFactory
import br.com.zup.beagle.android.view.custom.BeaglePageIndicatorView
import br.com.zup.beagle.android.widget.RootView
import br.com.zup.beagle.android.widget.WidgetView
import br.com.zup.beagle.annotation.RegisterWidget

/**
 *  The PageView component is a specialized container to hold pages (views) that will be displayed horizontally.
 *
 * @param selectedColor this is a string value and it must be filled as HEX (Hexadecimal)
 * @param unselectedColor this is a string value and it must be filled as HEX (Hexadecimal)
 * @param numberOfPages Numbers of pages that will contain.
 * @param currentPage Integer number that identifies that selected
 */
@RegisterWidget
class PageIndicator(
    val selectedColor: String,
    val unselectedColor: String,
    var numberOfPages: Int? = null,
    var currentPage: Bind<Int>? = null
) : WidgetView(), PageIndicatorComponent {

    constructor(
        selectedColor: String,
        unselectedColor: String,
        numberOfPages: Int? = null,
        currentPage: Int
    ) : this(selectedColor, unselectedColor, numberOfPages, valueOf(currentPage))

    @Transient
    private val viewFactory: ViewFactory = ViewFactory()

    @Transient
    private lateinit var pageIndicator: BeaglePageIndicatorView

    override fun buildView(rootView: RootView) = viewFactory.makePageIndicator(rootView.getContext()).apply {
        pageIndicator = this
        setSelectedColor(Color.parseColor(selectedColor))
        setUnselectedColor(Color.parseColor(unselectedColor))
        numberOfPages?.let {
            setCount(it)
        }
        currentPage?.let {
            observeBindChanges(rootView, this, it) { position ->
                position?.let {
                    onItemUpdated(position)
                }
            }
        }
    }


    override fun setCount(pages: Int) {
        pageIndicator.setCount(pages)
    }

    override fun onItemUpdated(newIndex: Int) {
        pageIndicator.setCurrentIndex(newIndex)
    }

    override fun initPageView(pageIndicatorOutput: PageIndicatorOutput) {}
}

File: iOS/Sources/Beagle/Sources/Components/ServerDrivenComponent/PageIndicator/PageIndicator.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/// The PageIndicator is a component to be used with PageView to sinalize the page selected.
public class PageIndicator: PageIndicatorComponent, AutoInitiable {
    
    /// Configures the color of the selected dot. Must be filled as HEX (Hexadecimal).
    public var selectedColor: String?
    
    /// Configures the color of the unselected dots. Must be filled as HEX (Hexadecimal).
    public var unselectedColor: String?
    
    /// Numbers of pages of the PageView.
    public var numberOfPages: Int?
    
    /// Integer number that identifies the page selected.
    public var currentPage: Expression<Int>?

// sourcery:inline:auto:PageIndicator.Init
    public init(
        selectedColor: String? = nil,
        unselectedColor: String? = nil,
        numberOfPages: Int? = nil,
        currentPage: Expression<Int>? = nil
    ) {
        self.selectedColor = selectedColor
        self.unselectedColor = unselectedColor
        self.numberOfPages = numberOfPages
        self.currentPage = currentPage
    }
// sourcery:end
}
